cmake/version.cmake

```diff
@@ -0,0 +1,54 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the Apache 2.0 License.
+
+# If possible, deduce project version from git environment.
+if(EXISTS ${CMAKE_CURRENT_SOURCE_DIR}/.git)
+  find_package(Git)
+
+  execute_process(
+    COMMAND git describe --tags
+    OUTPUT_VARIABLE "CCF_VERSION"
+    OUTPUT_STRIP_TRAILING_WHITESPACE
+  )
+
+  execute_process(
+    COMMAND "bash" "-c" "git describe --tags --abbrev=0 | tr -d v"
+    OUTPUT_VARIABLE "CCF_RELEASE_VERSION" OUTPUT_STRIP_TRAILING_WHITESPACE
+  )
+
+  # Check that git release tag follows semver
+  execute_process(
+    COMMAND
+      "bash" "-c"
+      "[[ ${CCF_RELEASE_VERSION} =~ ^([[:digit:]])+(\.([[:digit:]])+)*$ ]]"
+    RESULT_VARIABLE "TAG_IS_SEMVER"
+  )
+
+  if(NOT ${TAG_IS_SEMVER} STREQUAL "0")
+    message(
+      WARNING
+        "git tag \"${CCF_RELEASE_VERSION}\" does not follow semver. Defaulting to project version 0.0.0"
+    )
+    set(CCF_RELEASE_VERSION "0.0.0")
+  endif()
+else()
```

Comment:
Really minor edge case, but I think we could try the directory approach even if we're in a git directory, but the git approach failed? So

````
unset(CCF_RELEASE_VERSION)
if (is_git)
 // try via git, maybe calling set(CCF_RELEASE_VERSION ...)
endif()

if (! set(CCF_RELEASE_VERSION)) // This isn't the right syntax, but whatever
 // try via dirname
endif()

if (! set(CCF_RELEASE_VERSION))
  set(CCF_RELEASE_VERSION "0.0.0")
 message(WARNING "Couldn't auto-deduce version, defaulting to ${CCF_RELEASE_VERSION}")
endif()